### DISCUSSION TITLE
【TypeScript】

### DISCUSSION THREAD

**Discussion by swustzzh** [swustzzh/coder_books#32] [General]
![TS](https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/4d91762ec3c14c599e886e259a475102~tplv-k3u1fbpfcp-jj-mark:0:0:0:0:q75.avis)

**Comment by swlws** [comment]
在线练习：[https://www.typescriptlang.org/play](https://www.typescriptlang.org/play)

**Reply by swustzzh** [reply]
最好的练习地址是项目中实战，而不是一味的anyscript。

**Comment by swlws** [comment]
**组合 Key**

```ts
interface Left {
    Sender?: string
    Receive?: string
}

interface Right {
    Persion?: string
    Mobile?: string
}

type CombineKey = `${keyof Left}.${keyof Right}`

```

**Reply by swustzzh** [reply]
666，还可以这么玩，有啥场景嘛？来个看看

**Reply by swlws** [reply]
这是我最近碰见的场景：

两组表单，每一组表单都包含人名、手机号码等，一组为寄方信息、一组为收方信息。
每个表单项需要添加一个验证使用的字符串，比如 data-validate=“Sender.Person”
函数处理时，需要对参数中 data-validate做约束，就有了上面的声明

**Comment by swustzzh** [comment]
## 【TypeScript类型体操通关秘籍】
章节：1-3
- 理论很重要，提升认知，即看技术的角度
- 类型决定了变量的内存大小和可进行的操作
- 类型安全：保证对某种类型，只做符合该类型规范的操作
- 运行时做的检查叫动态类型检查，编译时做的检查叫静态类型检查
- 静态类型检查提升了写代码的难度，不仅要注意业务逻辑，还要注意类型逻辑，好处在于消除类型隐患，提高健壮性，适合大型项目
- 动态类型检查写代码灵活，但对类型不安全，适合简单场景
- JS是动态类型语言
- 静态类型编程语言从简单到复杂分为3类：
  - 简单类型系统：变量、函数、类等都可以生命类型，编译器做类型检查
  - 支持泛型的类型系统：声明时会把变化的类型声明成泛型，调用时确定类型
  - 支持类型编程的类型系统：对传入的类型做逻辑运算，产生新类型
- TS的类型是图灵完备，能描述各种可计算逻辑，js可以写的，TS类型都可以写

**Reply by swustzzh** [reply]
- 特殊的联合查询：
  - 条件类型中，左边是联合类型，会把每个元素单独传入做计算，右边不会
- 联合类型中的每个类型都是相互独立的，TS对其做了特殊处理，即遇到字符串、条件类型时，会把每个类型单独计入做计算，最后把每个类型的计算结果合并成联合类型
- any类型与任何类型的交叉都是any，也就是1&any=any
- never作为类型参数出现在条件类型左侧时，会直接返回never
- any作为类型参数出现在条件类型左侧时，会直接返回trueType和falseType的联合类型
- 元组类型也是数组类型，但length时数字字面量，而数组的length是number，可以用来判断元组类型
- 函数参数处会发生逆变，可以用来实现联合类型转交叉类型
- 可选索引的索引可能没有，那 Pick 出来的就可能是 {}，可以用来过滤可选索引，反过来也可以过滤非可选索引。
- 索引类型的索引为字符串字面量类型，而可索引签名不是，可以用这个特性过滤掉可索引签名。
- keyof 只能拿到 class 的 public 的索引，可以用来过滤出 public 的属性。
- 默认推导出来的不是字面量类型，加上 as const 可以推导出字面量类型，但带有 readonly 修饰，这样模式匹配的时候也得加上 readonly 才行。

**Reply by swustzzh** [reply]
TS顺口溜：

**模式匹配做提取，重新构造做变换。**
**递归复用做循环，数组长度做计数。**
**联合分散可简化，特殊特性要记清。**
**基础扎实套路熟，类型体操可通关。**

**Comment by swustzzh** [comment]
## 【TypeScript类型体操通关秘籍】
### 章节：5
#### 模式匹配做提取
- ts类型的模式匹配是通过extends对类型参数做匹配，结果保存到通过infer声明的局部类型变量里，如果匹配就能从该局部变量里，拿到提取的类型
 [例子](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgTgSwLYBkIDNgB4DKw5QQAewEAdgCYDOUV+CZA5gHxQC8UeBxplNABgBIA3gHIo4gD4SAOmSmzgogL4iG6CARwBXdOgRFl-KAH5YiVBmw69B1gC5O+ANwAoUJHPIASgkYALa3xCEnJqWnomVg4uEN5woWF1TSgABTgMA1UxCShpUTkFAqUjUy8kXwDsdMyiByc4Nw9oeGRcYJ4wmjpEKPZytEwsVoq-QPa4ZmY3d3BoAHEIYBHB4H6VqyxxKCgARgAmAGY9o52d0WnZz0Xli0rA9bux7G2T44Pjs4umuagbkcebVeHzeZwklyuCyWqQAhnAYUglpoqFgAGLaMgAYziXSg6KxwAQAHsyNE8RjsZ0+FAABQAOgZcMYVEcyQIAEE4IwAJTsVgYgDWZCJAHcyGVOYwoI4yBAAG6aGaQv5LfGY2HwxGkOA0Dg3DUIpE6rA0mEy7RIABGmgANFBLearZpeWxWHKiQgKBDmirgN4lto4GQYHM0RScdS1YSSWS1RHwvTGVyWVAYWQQABtAC6LtYbKg-uAgeDvzMheLIc8Mvlitcypuasr0D1S3LQabJtzUDIFutkxmPv9YAANjDMRAJvHupFGHbUXAiUgpxFerPYETlz0GCx+rEqQk1GQNAQavpDCJ54vSmZEqesiIYETSo4uAPfkPR+PYhwP2OJ6IYUtTEKFEO1RBAsCIAuIA)

模式匹配有点意思
